pennylane/plugins/default_qubit.py

```diff
@@ -213,7 +221,51 @@ def mat_vec_product(self, mat, vec, wires):
         perm = wires + unused_idxs
         inv_perm = np.argsort(perm)  # argsort gives inverse permutation
         state_multi_index = np.transpose(tdot, inv_perm)
-        return np.reshape(state_multi_index, 2 ** self.num_wires)
+
+        return state_multi_index
+
+    def mat_vec_product_einsum(self, mat, vec, wires):
+        r"""Apply multiplication of a matrix to subsystems of the quantum state.
+
+        This function uses einsum instead of tensordot. This approach is only
+        faster for single- and two-qubit gates.
+
+        Args:
+            mat (array): matrix to multiply
+            vec (array): state tensor to multiply
+            wires (Sequence[int]): target subsystems
+
+        Returns:
+            array: output vector after applying ``mat`` to input ``vec`` on specified subsystems
+        """
+        mat = np.reshape(mat, [2] * len(wires) * 2)
+
+        # Tensor indices of the quantum state
+        state_indices = ABC[: self.num_wires]
+
+        # Indices of the quantum state affected by this operation
+        affected_indices = "".join(ABC_ARRAY[wires].tolist())
+
+        # All affected indices will be summed over, so we need the same number of new indices
+        new_indices = ABC[self.num_wires : self.num_wires + len(wires)]
+
+        # The new indices of the state are given by the old ones with the affected indices
+        # replaced by the new_indices
+        new_state_indices = functools.reduce(
+            lambda old_string, idx_pair: old_string.replace(idx_pair[0], idx_pair[1]),
+            zip(affected_indices, new_indices),
+            state_indices,
+        )
```

Comment:
always fund to implement einsum "programatically" :laughing:

Reply:
I spent a long time trying to do the same thing with `tensordot`, and came to the conclusion it isn't possible :(

Reply:
Yeah, there is definitely a lack of "tensor" functions in numpy. Also the credit for this einsum mat vec multiplication goes to Josh!